### PR TITLE
Create Supervised Model

### PR DESCRIPTION
**Pull request by ben-heil**
This PR creates and tests the `PytorchSupervised` class. It turns out there are a lot of things required to train a neural network, so there are modifications in a number of places.

Changes:
- Tell git to ignore a secrets.yml file that stores things like API tokens
- Add instructions for installing neptune to the readme
- Update the environment file with new pytorch and neptune dependencies
- Add encoding logic to the `RefineBioLabeledExpression` class to allow training and adjust classes to return in samples x genes format
- Implement `PytorchSupervised` and its underlying model `ThreeLayerClassifier`
- Create a configuration file to keep track of how test models should be trained
- Write model tests
- Write util tests

### CODE REVIEW

File: saged/datasets.py [hunk 1]
```diff
@@ -694,6 +695,11 @@ def __init__(self,
         self.sample_to_label = sample_to_label
         self.sample_to_study = sample_to_study
 
+        encoder = preprocessing.LabelEncoder()
```
Comment (jjc2718) [inline saged/datasets.py]: Calling this `encoder` may get a little confusing if you're planning to generate data using an autoencoder in the future. Maybe `label_encoder` would be clearer?

File: saged/datasets.py [hunk 2]
```diff
@@ -342,9 +343,9 @@ def get_all_data(self) -> Tuple[np.array, np.array]:
 
         Returns
         -------
-        X: The gene expression data in a genes x samples array
+        X: The gene expression data in a samples x genes array
```
Comment (jjc2718) [inline saged/datasets.py:346]: Does the data from refine.bio and other sources come in genes x samples format? I can see it being a bit confusing that you're storing it internally as a genes x samples dataframe, then transposing it here before returning it.

Might be more straightforward to just store it internally as a samples x genes dataframe (and require that orientation as input to your constructor). That's just my opinion though, ultimately either will work fine so I'll leave it up to you.

If you really wanted to get into the weeds of this, I'm pretty sure Pandas dataframes are stored in column-major order by default ([see here for a bit of background](https://en.wikipedia.org/wiki/Row-_and_column-major_order) and [here for details about Pandas](https://stackoverflow.com/questions/49420272/how-to-make-pandas-dataframe-fortran-type-ordered), Fortran ordered == column-major), although it's surprisingly hard to find official Pandas documentation of this. This means it would probably be faster to have the dimension you plan to slice by most often as columns, and the other as rows. It seems to me like you're planning to slice by samples more often than genes, so that would support the orientation you have now (genes x samples)...but I have no idea how much this distinction will actually matter in practice.

File: saged/models.py [hunk 3]
```diff
@@ -134,3 +161,348 @@ def predict(self, dataset: np.array) -> np.array:
...
+    A wrapper class implementing the ExpressionModel API while remaining modular enough
+    to accept any supervised classifier implementing the nn.Module API
+    """
+    def __init__(self,
+                 config: dict,
+                 optimizer: torch.optim.Optimizer,
+                 loss_class: torch.nn.modules.loss._WeightedLoss,
+                 model_class: nn.Module) -> None:
+        """
+        Standard model init function. We use pass instead of raising a NotImplementedError
```
Comment (jjc2718) [inline saged/models.py]: Are you actually using `pass` anywhere in this class? Or is this just here for future reference?

File: saged/models.py [hunk 4]
```diff
@@ -134,3 +161,348 @@ def predict(self, dataset: np.array) -> np.array:
...
+
+    @classmethod
+    def load_model(classobject,
+                   checkpoint_path: str,
+                   config: dict,
+                   optimizer: torch.optim.Optimizer,
+                   loss_fn: torch.nn.modules.loss._WeightedLoss,
+                   model_class: nn.Module) -> "PytorchSupervised":
+        """
+        Read a pickeled model from a file and return it
```
Comment (jjc2718) [inline saged/models.py]:
```suggestion
        Read a pickled model from a file and return it
```

File: saged/utils.py [hunk 5]
```diff
@@ -281,3 +285,60 @@ def get_samples_in_studies(samples: List[str],
...
+    Returns
+    -------
+    The integer labels predicted by the model
+    """
+    return torch.argmax(model_output, dim=-1)
+
+
+def count_correct(outputs: torch.Tensor, labels: torch.Tensor) -> int:
+    """
+    Calculate the nubmer of correct predictions in the given batch
```
Comment (jjc2718) [inline saged/utils.py]:
```suggestion
    Calculate the number of correct predictions in the given batch
```

File: saged/models.py [hunk 6]
```diff
@@ -134,3 +161,348 @@ def predict(self, dataset: np.array) -> np.array:
...
+        return x
+
+
+class PytorchSupervised(ExpressionModel):
+    """
+    A wrapper class implementing the ExpressionModel API while remaining modular enough
+    to accept any supervised classifier implementing the nn.Module API
+    """
+    def __init__(self,
+                 config: dict,
```
Comment (jjc2718) [inline saged/models.py:284]: Do you know of a good way to document what needs to be in this config, and (if anything) what parameters are optional? I guess as it stands your code will throw a `KeyError` if (for example) the learning rate or seed is missing, but it might be good to have a list of what's necessary somewhere centralized.

Even just documentation pointing to the example config in `test/data/test_config.yml` might serve this purpose.

File: saged/models.py [hunk 7]
```diff
@@ -134,3 +161,348 @@ def predict(self, dataset: np.array) -> np.array:
...
+        self.model = model_class(model_config)
+        lr = float(model_config['lr'])
+        weight_decay = float(model_config.get('weight_decay', 0))
+        self.optimizer = optimizer(self.model.parameters(),
+                                   lr=lr,
+                                   weight_decay=weight_decay)
+
+        self.loss_class = loss_class
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (jjc2718) [inline saged/models.py]: Do you always want to run PyTorch models on the GPU if one is available, or is there ever a case where you might want to override this and run things on the CPU?

I know in the past I've noticed shallow models (e.g. linear/logistic regression) in PyTorch can at times run faster on the CPU, but if you're only planning on using this class for multi-layer NNs (which is what it looks like here) you might be fine.

**Review by jjc2718 — APPROVED** [review]
Looks good to me! Just a few things to think about.

File: saged/datasets.py (hunk 1 above)
Comment (ben-heil) [reply saged/datasets.py]: Good call

File: saged/datasets.py (hunk 2 above)
Comment (ben-heil) [reply saged/datasets.py:346]: That's definitely a good point. I currently have the internal representation stored as genes x samples because that's how they're stored in the compendium file. Since my most common internal operations are subsetting by sample that should be more efficient. Both sklearn and pytorch expect their inputs to be samples x features though, so I take the transpose when returning results to keep the data consistent with what the models are expecting

File: saged/models.py (hunk 3 above)
Comment (ben-heil) [reply saged/models.py]: Nope, good catch. That's a holdover from the ExpressionModel abstract class that does use `pass` for its init function

File: saged/models.py (hunk 6 above)
Comment (ben-heil) [reply saged/models.py:284]: That's a good question. I think it will make sense to write a custom parser function that validates the file, but I'll wait for a bit before doing so to let the config format solidify.

I think for now I'll add a reference in the README and add a TODO to the code

File: saged/models.py (hunk 7 above)
Comment (ben-heil) [reply saged/models.py]: That's a good point, it makes sense to put that in the config file too
